nstar/src/commands.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019 - 2020 ESRLabs
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use clap::{AppSettings, Result};
+use npk::manifest::Version;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+pub enum Northstar {
+    #[structopt(about = "List available containers", alias = "ls", alias = "list")]
+    Containers,
+    #[structopt(about = "List configured repositories", alias = "repos")]
+    Repositories,
+    #[structopt(about = "Start a container")]
+    Start {
+        #[structopt(help = "Name of the container")]
+        name: Option<String>,
+    },
+    #[structopt(about = "Stop a container")]
+    Stop {
+        #[structopt(help = "Name of the container")]
+        name: Option<String>,
+    },
+    #[structopt(about = "Install a container")]
+    Install {
+        #[structopt(help = "Path to the .npk file")]
+        npk: PathBuf,
+        #[structopt(help = "Target repository")]
+        repo_id: String,
+    },
+    #[structopt(about = "Uninstall a container")]
+    Uninstall {
+        #[structopt(help = "Name of the container")]
+        name: String,
+        #[structopt(help = "Container version")]
+        version: Version,
+    },
+    #[structopt(about = "Shutdown Northstar")]
+    Shutdown,
+}
+
+#[derive(StructOpt)]
+pub enum Prompt {
+    #[structopt(about = "Print help", alias = "h", alias = "?")]
+    Help,
+    #[structopt(about = "Quit", alias = "q", alias = "exit")]
+    Quit,
+}
+
+#[derive(StructOpt)]
+pub struct NstarOpt {
+    #[structopt(help = "Url to Northstar instance", default_value = "localhost:4200")]
+    pub host: String,
+    #[structopt(
+        short,
+        long,
+        help = "If present, print Northstar's response in JSON format"
+    )]
+    pub json: bool,
```

Comment:
The json option doesn't work in interactive mode yet. Either fix the CLI to conflict `-j` with a `command: None` or implement json for the interactive mode.

Reply:
does it make sense to have a json mode in interactive mode?

Reply:
I think yes. Image you're hacking a client and want to actually manually test sequences etc. But for sure this is minor.

Reply:
Do you mean that instead of having the pretty tables the json of each response is printed? I don't see how is that useful.

Reply:
Turns out that you can't set a top level flat like `-j` to require `command` to be `Some` with `clap`. An alternative would be to add the `-j` flag to every command which is not that nice. I resorted to do a manual check for that case after parsing the arguments.

Reply:
It's super easy to have json also in the "interactive mode". I consider this as useful for manual check during the development of a client. We get this feature for free and don't have to mess with the `-j` flag because it can be global.